tools/apigee-proxy-modifier-validator/README.md

```diff
@@ -0,0 +1,170 @@
+
+## Purpose
+
+During migration from Apigee Edge to Apigee X/Hybrid, certain policy configurations, endpoint settings, or target URLs might need adjustments for compatibility. This script automates the process of:
+
+1.  Extracting an API proxy bundle (`.zip`).
+2.  Parsing XML files within `apiproxy/policies/`, `apiproxy/proxies/`, and `apiproxy/targets/`.
+3.  Modifying XML elements or attributes based on rules defined in a TOML file (using `lxml` and **XPath**). Supported operations include Set/Replace, Prefix, Suffix, and Conditional Prefix.
+4.  Preserving original XML formatting (whitespace, comments, etc.) as much as possible during modification thanks to `lxml`.
+5.  Generating detailed `diff` output only for files with actual content changes.
+6.  Re-zipping the modified bundle structure.
+7.  Optionally validating the modified bundle using the Apigee API (inferring the proxy name from the filename).
+8.  Processing multiple bundles from an input directory to an output directory.
+9.  Generating a comprehensive **Markdown report file** (`.md`) containing:
+    *   Overall processing summary counts.
+    *   A summary table of all processed bundles.
+    *   Detailed status per bundle, including validation results and error snippets.
+    *   Line-by-line diffs (using `difflib`) for all files that were modified.
+
+## Features
+
+*   Modifies XML files (Policies, ProxyEndpoints, TargetEndpoints) within Apigee proxy bundles.
+*   Uses the **`lxml` library** for robust XML parsing and modification, aiming to preserve formatting.
+*   Uses a TOML file for defining modification rules using **XPath expressions**.
+*   Supports multiple modification operations: Set/Replace (with wildcard 'from'), Prefix, Suffix, Conditional Prefix.
+*   Processes all `.zip` files in a specified input directory to an output directory.
+*   Generates accurate `diff` output highlighting content changes, minimizing formatting noise.
+*   Optionally validates modified bundles against the Apigee API using Google ADC.
+*   Infers the proxy name for validation from the bundle filename (requires consistent naming).
+*   Generates a detailed **Markdown report file**.
+*   Supports overwriting existing output files (`--overwrite`).
+*   Verbose console logging option (`-v`) for real-time progress and debugging.
+
+## Prerequisites
+
+1.  **Python:** Python 3.8 or higher recommended.
+2.  **pip:** Python package installer.
+3.  **Google Cloud SDK:** (`gcloud` command-line tool) installed and authenticated.
+4.  **Application Default Credentials (ADC):** Configured for Google Cloud authentication. Run `gcloud auth application-default login` in your terminal and authenticate with a Google account that has permissions to validate Apigee APIs (e.g., Apigee API Admin role or custom role with `apigee.organizations.apis.validate`).
```

Comment:
I don't believe `apigee.organizations.apis.validate` is a valid [permission](https://cloud.google.com/iam/docs/roles-permissions/apigee#apigee-permissions).

Reply:
fixed it